### PR TITLE
[no-release-notes] New format encodings for `Year`, `Date` and `Time`

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**

### CODE REVIEW

**Review by Hydrocharged — APPROVED** [review]
Overall I think this looks fine. I think most of my feedback can be resolved by moving some of the logic to the `TypeInfo` package, but that's a separate PR, so feel free to ignore it for now.

File: go/store/prolly/shim.go [hunk 1]
```diff
@@ -129,29 +128,26 @@ func ConvertToKeylessIndex(m Map) Map {
...
+	case query.Type_BIT:
+		return val.Uint64Enc
+	case query.Type_BLOB:
+		return val.ByteStringEnc
+	case query.Type_TEXT:
+		return val.StringEnc
+	case query.Type_ENUM:
+		return val.StringEnc
+	case query.Type_SET:
+		return val.StringEnc
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/store/prolly/shim.go:148]: `SET` and `ENUM` are still using temporary encodings, I could use some help picking an encoding format for them.

I moved them to the top of this function to reflect their temporary status.

File: go/store/val/codec.go [hunk 2]
```diff
@@ -359,18 +365,83 @@ func compareFloat64(l, r float64) int {
...
+// adapted from:
+// https://dev.mysql.com/doc/internals/en/date-and-time-data-type-representation.html
+const (
+	yearShift  uint32 = 16
+	monthShift uint32 = 8
+	monthMask  uint32 = 255 << monthShift
+	dayMask    uint32 = 255
+)
+
+func readDate(val []byte) (date time.Time) {
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/store/val/codec.go:398]: I was also somewhat conflicted. Ultimately MySQL provides these discrete data types to give database developers more options/control for their schema, so I think we should provide as much flexibility.

File: go/store/prolly/shim.go [hunk 3]
```diff
@@ -129,29 +128,26 @@ func ConvertToKeylessIndex(m Map) Map {
 	}
 }
 
-// todo(andy): move this to typeinfo
-func encodingFromTypeInfo(ti typeinfo.TypeInfo) val.Encoding {
-	if ti.GetTypeIdentifier() == typeinfo.UuidTypeIdentifier {
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/store/prolly/shim.go:134]: `Uuid` is a type in Noms, but not in SQL. We don't have any plans to continue this type in the new format

**Pull request merged**